deep_qa/data/fields/field.py

```diff
@@ -0,0 +1,117 @@
+tokens.  If the field has text, it is still strings, but it's lists of strings instead of a single
+string.  These fields are used to compute vocabularies.  At this point we know what tokens there
+are in the data, and we can count them to set a fixed vocabulary for things like embedding layers
+in a model.
+
+Tokenized fields become arrays using various :class:`TokenRepresentation` objects.
+
+Indexed fields
+~~~~~~~~~~~~~~
+Indexed fields are no longer raw - there are no strings left, only numeric types that can be used
+directly in data arrays.  They still aren't quite yet ready to be used for training, however,
+because the sizes of the arrays are not consistent across instances.  Before a batch of data can be
+passed to the model for training or evaluation, these fields need to be padded.  This is done by a
+:class:`~deep_qa.data.data_generator.DataGenerator`.
+
+Padded fields
+~~~~~~~~~~~~~
+Padded fields are numpy arrays that are passed directly to training / evaluation loops.  We don't
+actually instantiate any PaddedField objects in this code; ``IndexedFields`` just return arrays
+after padding.  We just mention this here as the final endpoint of ``RawFields``.
+"""
+from typing import Dict, List
+
+import numpy
+
+from ..vocabulary import Vocabulary
+
+# Because of how the dependencies work between these objects, we're defining them in reverse order,
+# so that TokenizedField can reference IndexedField in its return types, and RawField can reference
+# TokenizedField.
+
+class IndexedField:
+    """
+    An ``IndexedField`` is a field that is almost ready to be put into a data array for use in a
+    model.  The only thing missing is padding.  The methods on this object allow you to get padding
+    lengths, so you can group things into batches of similar size, and then actually pad the
+    field into a fixed-size array that can be used in a batch of instances.
+    """
+    def get_padding_lengths(self) -> Dict[str, int]:
+        """
+        If there are things in this field that need padding, note them here.  In order to pad a
+        batch of instance, we get all of the lengths from the batch, take the max, and pad
+        everything to that length (or use a pre-specified maximum length).  The return value is a
+        dictionary mapping keys to lengths, like {'num_tokens': 13}.
+        """
+        raise NotImplementedError
+
+    def pad(self, padding_lengths: Dict[str, int]) -> List[numpy.array]:
+        """
+        Given a set of specified padding lengths, actually pad the data in this field and return a
+        numpy array of the correct shape.  This actually returns a list instead of a single array,
+        in case there are several related arrays for this field (e.g., a ``TextField`` might have a
+        word array and a characters-per-word array).
+        """
+        raise NotImplementedError
+
+
+class TokenizedField:
```

Comment:
Maybe?  I debated about that.  It probably makes sense, but it also allows for easy errors.  E.g., if you're reading in an image, you can probably go straight to `IndexedField`, but then the dataset / instance class has to be sure that it's not calling `count_vocab_items` when it's iterating over the fields.

So, how about this:

The instances can take any of `RawField`, `TokenizedField`, or `IndexedField` as their input.  The first step will be to iterate over all of the fields, and if they're `Raw` (or maybe just `Untokenized`), we tokenize them.  Then we iterate over all of the fields again, and count vocab items on all of the `TokenizedFields`.  Then we iterate over all of them again, and convert `TokenizedFields` to `IndexedFields`.  It's a bunch of type checks on the fields, but it's not too bad, just checking the field's super class.

This way, when we have a `LabelField`, or a `TagField`, they start out as `TokenizedFields`, and have no `RawField` version - the `Dataset` that's constructing the instances doesn't have to think too hard about what level the field is getting constructed at, because there will only be one public class for each field type.  Does this make sense?